### PR TITLE
Create CLI Issue 126

### PR DESCRIPTION
**Pull request by bdegley4789**
As I said I would create this pull request today for you to review. Let me know if I am making any mistakes. 
I made the changes to setup.py you had mentioned in Issue #126 but I'm still a little bit confused on what I should include in the main functions of pycoal.cli.mineral:main, pycoal.cli.mining:main, and pycoal.cli.environment:main
I can also work on testing this CLI tomorrow. I thought [this ](https://www.npmjs.com/package/test-cli)would be a good start for that.  Unless, this is already good as is.

### CODE REVIEW

File: pycoal/cli/environment.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
+# Copyright (C) 2017 COAL Developers
```
Comment (lewismc) [inline pycoal/cli/environment.py]: It is not 2017

File: pycoal/cli/environment.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+# of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public
+# License along with this program; if not, write to the Free
+# Software Foundation, Inc., 51 Franklin Street, Fifth
+# Floor, Boston, MA 02110-1301, USA.
+# encoding: utf-8
+'''
+example_environment -- an example script which demonstrates COAL
```
Comment (lewismc) [inline pycoal/cli/environment.py]: This is not an example, it is the CLI for the toolkit.

File: pycoal/cli/environment.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+environmental classification
+example_environment provides a CLI which demonstrates how the 
+COAL Environmental Classification API finds pixels in a mining
+classified image that are within a certain number of meters
+from features in a vector layer such as flow lines in the
+National Hydrography Dataset (NHD).
+More reading an this example can be seen at 
+https://capstone-coal.github.io/docs#usage
+@author:     COAL Developers
+@copyright:  2017 COAL Developers. All rights reserved.
```
Comment (lewismc) [inline pycoal/cli/environment.py]: Again it is not 2017

File: pycoal/cli/environment.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import environment
+
+__all__ = []
+
+DEBUG = 1
+TESTRUN = 0
+PROFILE = 0
+
+from pycoal.write import Writer
+from pycoal.read_netcdf import NetCDFReader as Reader
```
Comment (lewismc) [inline pycoal/cli/environment.py]: The code above has been copied from pycovjson, the ```Writer``` and ```NetCDFReader``` Objects do not exist in the pycoal project!

File: pycoal/cli/environment.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    if argv is None:
+        argv = sys.argv
+    else:
+        sys.argv.extend(argv)
+
+    program_name = os.path.basename(sys.argv[0])
+    program_shortdesc = __import__('__main__').__doc__.split("\n")[1]
+    program_license = '''%s
+  VERSION %s
+  Copyright 2017 COAL Developers. All rights reserved.
```
Comment (lewismc) [inline pycoal/cli/environment.py]: It is not 2017

File: pycoal/cli/environment.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    try:
+        # Setup argument parser
+        parser = ArgumentParser(
+            description=program_license,
+            formatter_class=RawDescriptionHelpFormatter)
+        parser.add_argument(
+            "-m", "--mining",
+            dest="mining_filename",
+            default='ang20150420t182050_corr_v1e_img_class_mining.hdr',
+            help="Input mining classified file to be processed [default: ang20150420t182050_corr_v1e_img_class_mining.hdr]")
```
Comment (lewismc) [inline pycoal/cli/environment.py]: There should be no default files... as this is not an example, the user should provide these files.

File: pycoal/cli/environment.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        parser.add_argument(
+            "-hy", "--hydrography",
+            dest="vector_filename",
+            default='Shape/NHDFlowline.shp',
+            help="Path to hydrography data [default: Shape/NHDFlowline.shp]")
```
Comment (lewismc) [inline pycoal/cli/environment.py]: There should be no Shape file, as this is not an example, the user should provide these.

File: pycoal/cli/environment.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        parser.add_argument(
+            "-e", "--environment",
+            dest="correlation_filename",
+            default='ang20150420t182050_corr_v1e_img_class_mining_NHDFlowline_correlation.hdr',
+            help="Output environmental correlation image [default: ang20150420t182050_corr_v1e_img_class_mining_NHDFlowline_correlation.hdr]")
```
Comment (lewismc) [inline pycoal/cli/environment.py]: There should be no correlation file name chosen by default, as this is not an example, the user should provide these.

File: pycoal/cli/environment.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+        # Process arguments
+        args = parser.parse_args(
+            ['-m', 'ang20150420t182050_corr_v1e_img_class_mining.hdr',
+            '-hy', 'Shape/NHDFlowline.shp',
+            '-e', 'ang20150420t182050_corr_v1e_img_class_mining_NHDFlowline_correlation.hdr'])
```
Comment (lewismc) [inline pycoal/cli/environment.py]: Again, there should be no example files, as this is not an example, the user should provide these.

File: pycoal/cli/environment.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+        mining_filename = args.mining_filename
+        vector_filename = args.vector_filename
+        correlation_filename = args.correlation_filename
+
+        run_environment(mining_filename, vector_filename, correlation_filename)
```
Comment (lewismc) [inline pycoal/cli/environment.py]: Where are you getting this function from? It does not exist right!

**Review by lewismc — CHANGES_REQUESTED** [review]
@bdegley4789 I've reviewed your proposed addition ```pycoal/cli/environment.py```, I've highlighted several areas where things are not quite right. Take a look at my comments. They are the same for the other proposed additions as well.
Did you try running this stuff before you submitted the pull request?

**Comment by lewismc** [comment]
@bdegley4789 there is absolutely no way you can have installed this and attempted to run locally. When I attempt to install I get the following
```
Error in sitecustomize; set PYTHONVERBOSE for traceback:
KeyError: 'PYTHONPATH'
Traceback (most recent call last):
  File "setup.py", line 48, in <module>
    _entry_points = entry_points
NameError: name 'entry_points' is not defined
```
The syntax for the ```entry_points``` entry in ```setup.py``` is incorrect!

File: setup.py [hunk 11]
```diff
@@ -78,4 +79,5 @@ def _read(*rnames):
     url=_url,
     version=_version,
     zip_safe=_zip_safe,
+    _entry_points = {'console_scripts' : ['pycoal-mineral = pycoal.cli.mineral:main', 'pycoal-mining = pycoal.cli.mining:main', 'pycoal-environment = pycoal.cli.environment:main']}
```
Comment (lewismc) [inline setup.py]: The syntax should be ```entry_points``` not ```_entry_points```, please correct and update pull request

**Comment by bdegley4789** [comment]
I created the pull request more so you could review it and clear up my confusions on how to create the cli. I didn't try to run the cli files the first time. In hind sight, I probably should have just asked you to look at the branch I created with changes I committed instead of creating a pull request

**Comment by bdegley4789** [comment]
I made all the changes you had requested. There are still a few more things on the CLI I would like to work on tomorrow. We could accept the pull request now though since it would fix the Travis CI build which is currently broken

**Comment by lewismc** [comment]
Hi @bdegley4789 this is a substantial improvement thank you. The command line tools are successfully created now and can be invoked from the terminal. 
There are lingering issues however, please see below and fix
```
lmcgibbn@LMC-056430 /usr/local/pycoal(Issue-126) $ pycoal-mineral
Error in sitecustomize; set PYTHONVERBOSE for traceback:
KeyError: 'PYTHONPATH'
Traceback (most recent call last):
  File "/Users/lmcgibbn/miniconda3/bin/pycoal-mineral", line 11, in <module>
    load_entry_point('pycoal==0.5.2', 'console_scripts', 'pycoal-mineral')()
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 561, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2631, in load_entry_point
    return ep.load()
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2291, in load
    return self.resolve()
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2297, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/Users/lmcgibbn/miniconda3/lib/python3.6/site-packages/pycoal-0.5.2-py3.6.egg/pycoal/cli/mineral.py", line 36, in <module>
    import mineral
ModuleNotFoundError: No module named 'mineral'
```
```
lmcgibbn@LMC-056430 /usr/local/pycoal(Issue-126) $ pycoal-mining
Error in sitecustomize; set PYTHONVERBOSE for traceback:
KeyError: 'PYTHONPATH'
Traceback (most recent call last):
  File "/Users/lmcgibbn/miniconda3/bin/pycoal-mining", line 11, in <module>
    load_entry_point('pycoal==0.5.2', 'console_scripts', 'pycoal-mining')()
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 561, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2631, in load_entry_point
    return ep.load()
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2291, in load
    return self.resolve()
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2297, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/Users/lmcgibbn/miniconda3/lib/python3.6/site-packages/pycoal-0.5.2-py3.6.egg/pycoal/cli/mining.py", line 48
    if argv is None:
    ^
IndentationError: unexpected indent
```
```
lmcgibbn@LMC-056430 /usr/local/pycoal(Issue-126) $ pycoal-environment
Error in sitecustomize; set PYTHONVERBOSE for traceback:
KeyError: 'PYTHONPATH'
Traceback (most recent call last):
  File "/Users/lmcgibbn/miniconda3/bin/pycoal-environment", line 11, in <module>
    load_entry_point('pycoal==0.5.2', 'console_scripts', 'pycoal-environment')()
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 561, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2631, in load_entry_point
    return ep.load()
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2291, in load
    return self.resolve()
  File "/usr/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2297, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/Users/lmcgibbn/miniconda3/lib/python3.6/site-packages/pycoal-0.5.2-py3.6.egg/pycoal/cli/environment.py", line 37, in <module>
    import mineral
ModuleNotFoundError: No module named 'mineral'
```
Please fix and update the PR, in the future it would be advantageous to test your code locally before pushing to the remote branch. Thanks this is looking much better.

**Comment by bdegley4789** [comment]
Hey @lewismc I have fixed the errors you encountered. 

I'm stuck on passing the files as arguments. I tried a lot of different ways to pass arguments and I keep getting the same expected one argument error. I'm probably making a simple mistake here. 

I was able to successfully run the code though and generate correlation images by just typing the file names within the script. So, I think everything works here if you can correctly pass the file names as arguments

`10-251-4-55:cli bryceegley$ python mining.py -mi ang20150420t182050_corr_v1e_img_class.hdr -mo ang20150420t182050_corr_v1e_img_class_mining.hdr
usage: mining.py [-h] [-mi INPUT] [-mo OUTPUT]
mining.py: error: argument -mi/--mineral_input: expected one argument`

**Comment by lewismc** [comment]
@bdegley4789 I've pushed an update to your branch, please scope out the changes I've made to pycoal/cli/mineral.py
This will simplify things.
Please just use this logic for the remaining tools. Update your PR when you have time.
Thank you

**Comment by bdegley4789** [comment]
I'm not sure what's wrong with Travis-Ci. I made the changes to mining and environment.py in the cli modeled off the changes you @lewismc did to mineral.py and I am able to successfully run all the scripts using the filenames as arguments
`10-249-231-199:cli bryceegley$ python3 mining.py -mi ang20150420t182050_corr_v1e_img_class.hdr -mo ang20150420t182050_corr_v1e_img_class_mining.hdr

Saving /Users/bryceegley/Desktop/pycoal/pycoal/cli/ang20150420t182050_corr_v1e_img_class_mining.img
 10-249-231-199:cli bryceegley$ `

**Comment by bdegley4789** [comment]
Ok, I fixed the problem with travis-ci.
We need to have 
`import sys
import pycoal
sys.path.insert(0, '../')
import mining
from mining import MiningClassification`
In every file where we import pycoal classification scripts. Or change the path if the file is in a different location

**Comment by lewismc** [comment]
Hi @bdegley4789 with the changes I just pushed you can now execute from the command line as follows
```
lmcgibbn@LMC-056430 /usr/local/pycoal(Issue-126) $ pycoal-mineral -h
Error in sitecustomize; set PYTHONVERBOSE for traceback:
KeyError: 'PYTHONPATH'
usage: pycoal-mineral [-h] [-i IMAGE] [-s SLIB] [-r RGB_FILENAME]
                      [-c CLASSIFIED_FILENAME]

pycoal-mineral -- a CLI for COAL mineral classification

pycoal-mineral provides a CLI which demonstrates how the COAL Mineral Classification
API provides methods for generating visible-light and mineral classified images.
Mineral classification can take hours to days depending on the size of the spectral
library and the available computing resources, so running a script in the background
is recommended. More reading an this example can be seen at
https://capstone-coal.github.io/docs#usage

@author:     COAL Developers

@copyright:  2018 COAL Developers. All rights reserved.

@license:    GNU General Public License version 2

@contact:    coal-capstone@googlegroups.com

optional arguments:
  -h, --help            show this help message and exit
  -i IMAGE, --image IMAGE
                        Input file to be processed
  -s SLIB, --slib SLIB  Spectral Library filename
  -r RGB_FILENAME, --rgb_filename RGB_FILENAME
                        RGB File Name
  -c CLASSIFIED_FILENAME, --classified_filename CLASSIFIED_FILENAME
                        Classified File Name
```
```
lmcgibbn@LMC-056430 /usr/local/pycoal(Issue-126) $ pycoal-mining -h
Error in sitecustomize; set PYTHONVERBOSE for traceback:
KeyError: 'PYTHONPATH'
usage: pycoal-mining [-h] [-mi INPUT] [-mo OUTPUT]

pycoal-mining -- a CLI for COAL mining classification

pycoal-mining provides a CLI which demonstrates how the COAL Mining Classification
API provides methods for generating mining classified images.
Mining classification runtime depends largely on the size of the spectral
library and the available computing resources. More reading an this example can be seen at
https://capstone-coal.github.io/docs#usage

@author:     COAL Developers

@copyright:  2018 COAL Developers. All rights reserved.

@license:    GNU General Public License version 2

@contact:    coal-capstone@googlegroups.com

optional arguments:
  -h, --help            show this help message and exit
  -mi INPUT, --mineral_input INPUT
                        Input classified mineral file to be processed
  -mo OUTPUT, --mining_output OUTPUT
                        Output mining classified image filename
```
```
lmcgibbn@LMC-056430 /usr/local/pycoal(Issue-126) $ pycoal-environment -h
Error in sitecustomize; set PYTHONVERBOSE for traceback:
KeyError: 'PYTHONPATH'
usage: pycoal-environment [-h] [-m MINING_FILENAME] [-hy VECTOR_FILENAME]
                          [-e CORRELATION_FILENAME]
pycoal-environment -- a CLI for COAL environment classification

pycoal-environment provides a CLI which demonstrates how the COAL Environment Classification
API provides methods for generating environment classified images.
Environment classification runtime depends largely on the size of the spectral
library and the available computing resources. More reading an this example can be seen at
https://capstone-coal.github.io/docs#usage

@author:     COAL Developers

@copyright:  2018 COAL Developers. All rights reserved.

@license:    GNU General Public License version 2

@contact:    coal-capstone@googlegroups.com

optional arguments:
  -h, --help            show this help message and exit
  -m MINING_FILENAME, --mining MINING_FILENAME
                        Input mining classified file to be processed
  -hy VECTOR_FILENAME, --hydrography VECTOR_FILENAME
                        Path to hydrography data
  -e CORRELATION_FILENAME, --environment CORRELATION_FILENAME
                        Output environmental correlation image
```
Thank you for your work on this.

**Pull request merged**